core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts

```
@@ -0,0 +1,249 @@
+      this.ruleFactory.createFromBackendDict({
+        filters: [{
+          type: PlatformParameterFilterType.ServerMode,
+          conditions: [['=', ServerMode.Dev.toString()]]
+        }],
+        value_when_matched: false
+      })
+    );
+  }
+
+  addNewFilter(rule: PlatformParameterRule): void {
+    rule.filters.push(
+      this.filterFactory.createFromBackendDict({
+        type: PlatformParameterFilterType.ServerMode,
+        conditions: []
+      })
+    );
+  }
+
+  addNewCondition(filter: PlatformParameterFilter): void {
+    const context = this.filterTypeToContext[filter.type];
+    filter.conditions.push([
+      context.operators[0],
+      context.options ? context.options[0] : ''
+    ]);
+  }
+
+  removeRule(feature: PlatformParameter, ruleIndex: number): void {
+    feature.rules.splice(ruleIndex, 1);
+  }
+
+  removeFilter(rule: PlatformParameterRule, filterIndex: number): void {
+    rule.filters.splice(filterIndex, 1);
+  }
+
+  removeCondition(
+      filter: PlatformParameterFilter, conditionIndex: number): void {
+    filter.conditions.splice(conditionIndex, 1);
+  }
+
+  moveRuleUp(feature: PlatformParameter, ruleIndex: number): void {
+    const rule = feature.rules[ruleIndex];
+    this.removeRule(feature, ruleIndex);
+    feature.rules.splice(ruleIndex - 1, 0, rule);
+  }
+
+  moveRuleDown(feature: PlatformParameter, ruleIndex: number): void {
+    const rule = feature.rules[ruleIndex];
+    this.removeRule(feature, ruleIndex);
+    feature.rules.splice(ruleIndex + 1, 0, rule);
+  }
+
+  async updateFeatureRulesAsync(feature: PlatformParameter): Promise<void> {
+    if (this.adminTaskManager.isTaskRunning()) {
+      return;
+    }
+    const commitMessage = this.windowRef.nativeWindow.prompt(
+      'This action is irreversible. If you insist to proceed, please enter ' +
+      'the commit message for the update',
+      `Update feature '${feature.name}'.`
+    );
+    if (commitMessage === null) {
+      return;
```

Suggest setting a status message in this case rather than silently failing.

---

It's not a fail, the commitMessage is null when the user clicks the cancel button on the browser prompt, I think it's intuitive to silently abort the update.

---

Ah, to be clear I was considering the case when the user gets the confirmation dialog & doesn't fill in a commit message, then clicks confirm. Won't that look the same as cancelling with the current logic?

---

As long as the user clicks the confirm button on the dialog, the `commitMessage` is a string, which may be an empty string if the user leave the message empty. 
Given that the user clicks the confirm button, I think we should proceed the update, and we don't have a strong constraint such that the commit message cannot be empty.

---

Ah, I see. Okay, sounds reasonable.